File: _drafts/old/2026-09/semi.py
from dataclasses import dataclass
from collections import Counter, deque
from typing import ClassVar, Optional


class MS(tuple):
    # multiset
    def __new__(cls, xs=()):
        return super().__new__(cls, sorted(xs))

    def __lt__(self, other):
        return (len(self), tuple(self)) < (len(other), tuple(other))


@dataclass(frozen=True, slots=True, init=False)
class Semi:
    outer_order: ClassVar[str] = "multilex"
    # Store coefficients instead of repeating monomials.  The old representation
    # made 16*x a tuple containing sixteen copies of ("x",).
    terms: tuple[tuple[MS[str], int], ...]

    def __init__(self, monoms=()):
        counts = Counter(MS(m) for m in monoms)
        object.__setattr__(self, "terms", tuple(sorted(counts.items())))

    @classmethod
    def from_counter(cls, counts: Counter) -> "Semi":
        obj = object.__new__(cls)
        object.__setattr__(
            obj,
            "terms",
            tuple(sorted((MS(m), n) for m, n in counts.items() if n > 0)),
        )
        return obj

    @property
    def monoms(self) -> MS[MS[str]]:
        # Compatibility for notebook exploration.  Core operations use `terms`
        # and never expand coefficients this way.
        return MS(m for m, n in self.terms for _ in range(n))

    @staticmethod
    def _lex_runs(a, b, reverse=False) -> bool:
        """Compare expanded monomial sequences without expanding coefficients."""
        aa = tuple(reversed(a)) if reverse else a
        bb = tuple(reversed(b)) if reverse else b
        i = j = 0
        ca = cb = 0
        while i < len(aa) and j < len(bb):
            ma, na = aa[i]
            mb, nb = bb[j]
            if ma != mb:
                return ma < mb
            if ca == 0:
                ca = na
            if cb == 0:
                cb = nb
            used = min(ca, cb)
            ca -= used
            cb -= used
            if ca == 0:
                i += 1
            if cb == 0:
                j += 1
        return i == len(aa) and j != len(bb)

    @staticmethod
    def lit(name) -> "Semi":
        return Semi([[name]])

    @staticmethod
    def of_int(n: int) -> "Semi":
        assert n >= 0
        return Semi.from_counter(Counter({MS(): n}))

    def __add__(self, other: "Semi") -> "Semi":
        if isinstance(other, int):
            other = Semi.of_int(other)
        counts = Counter(dict(self.terms))
        counts.update(dict(other.terms))
        return Semi.from_counter(counts)

    def __radd__(self, other: int) -> "Semi":
        return Semi.of_int(other) + self

    def __pow__(self, n: int) -> "Semi":
        assert n >= 0
        if n == 0:
            return Semi.of_int(1)
        result = self
        for _ in range(n - 1):
            result *= self
        return result

    def __mul__(self, other: "Semi") -> "Semi":
        counts = Counter()
        for m1, n1 in self.terms:
            for m2, n2 in other.terms:
                counts[MS(m1 + m2)] += n1 * n2
        return Semi.from_counter(counts)

    def __rmul__(self, other: int) -> "Semi":
        return Semi.of_int(other) * self

    def __repr__(self) -> str:
        pieces = []
        for m, n in self.terms:
            term = " ".join(m) if m else "1"
            pieces.extend([term] * n)
        return " + ".join(pieces)

    def __lt__(self, other: "Semi") -> bool:
        if Semi.outer_order == "multilex":
            return self._lex_runs(self.terms, other.terms, reverse=True)
        return self._lex_runs(self.terms, other.terms)

    def __sub__(self, other: "Semi") -> Optional["Semi"]:
        counts = Counter(dict(self.terms))
        for m, n in other.terms:
            if counts[m] < n:
                return None
            counts[m] -= n
        return Semi.from_counter(counts)

    def divrem(self, other: "Semi") -> tuple["Semi", "Semi"]:
        # returns largest q such that self = q*other + r
        assert isinstance(other, Semi)
        if not other.terms:
            raise ValueError("division by zero")
        q = Counter()
        r = Counter(dict(self.terms))
        lm = other.terms[-1][0]
        i = len(r) - 1
        while i >= 0:
            rms = sorted(m for m, n in r.items() if n > 0)
            if i >= len(rms):
                i = len(rms) - 1
            if i < 0:
                break
            rm = rms[i]
            qm = list(rm)
            for x in lm:
                if x not in qm:
                    break
                qm.remove(x)
            else:
                qm = MS(qm)
                shifted = [(MS(qm + m), n) for m, n in other.terms]
                copies = min(r[m] // n for m, n in shifted)
                if copies:
                    q[qm] += copies
                    for m, n in shifted:
                        r[m] -= copies * n
                    i = len(r)
                    continue
            i -= 1
        return Semi.from_counter(q), Semi.from_counter(r)

    def overlaps(self, other) -> list["Semi"]:
        # return all nontrivial overlaps of self and other
        # such that ov = q1 * self + r1 and ov = q2 * other + r2
        # ov is less that lm(self) * lm(other) ?
        res = []
        seen = set()
        for (q, r), _ in self.overlaps_qr(other):
            ov = q * self + r
            if ov not in seen:
                seen.add(ov)
                res.append(ov)
        return res

    def overlaps_qr(
        self, other
    ) -> list[tuple[tuple["Semi", "Semi"], tuple["Semi", "Semi"]]]:
        # Polynomial contexts split additively, so align one monomial occurrence.
        res = []
        for m1, _ in self.terms:
            for m2, _ in other.terms:
                cm = Counter(m1) | Counter(m2)
                qm1 = MS((cm - Counter(m1)).elements())
                qm2 = MS((cm - Counter(m2)).elements())
                q1, q2 = Semi([qm1]), Semi([qm2])
                t1, t2 = q1 * self, q2 * other
                c1, c2 = Counter(dict(t1.terms)), Counter(dict(t2.terms))
                ov = Semi.from_counter(c1 | c2)
                qr = ((q1, ov - t1), (q2, ov - t2))
                if qr not in res:
                    res.append(qr)
        return res


# hypothesis tests to check supposed properties


@dataclass(frozen=True, slots=True)
class Eq:  # It is itself a multiset
    lhs: Semi
    rhs: Semi


type Rewrites = list["Rewrite"]


def reduce(semi: Semi, rewrites: Rewrites) -> Semi:
    while True:
        for rw in rewrites:
            q, r = semi.divrem(rw.lhs)
            res = q * rw.rhs + r
            if res != semi:
                semi = res
                break
        else:
            return semi


@dataclass(slots=True)
class Rewrite:  # It is itself a multiset
    lhs: Semi
    rhs: Semi
    mark: bool = False

    def __init__(self, lhs: Semi, rhs: Semi, mark: bool = False):
        if lhs < rhs:
            lhs, rhs = rhs, lhs
        self.lhs, self.rhs, self.mark = lhs, rhs, mark

    def __str__(self) -> str:
        return f"{self.lhs} -> {self.rhs}"


def naive_complete(eqs: list[Eq]) -> list[Rewrite]:
    # naive completion algorithm
    pending = list(eqs)
    rws = []
    # in loop, reduce equations and add reduced oreitned form to rewrites
    while pending:
        eq = pending.pop(0)
        lhs, rhs = reduce(eq.lhs, rws), reduce(eq.rhs, rws)
        if lhs == rhs:
            continue
        rw = Rewrite(lhs, rhs)
        if rw in rws:
            continue
        # add all overlaps of rewrite lhs to equations
        rules = rws + [rw]
        for rw1 in rules:
            for (q, r), (q1, r1) in rw.lhs.overlaps_qr(rw1.lhs):
                lhs = reduce(q * rw.rhs + r, rules)
                rhs = reduce(q1 * rw1.rhs + r1, rules)
                eq = Eq(lhs, rhs)
                if lhs != rhs and eq not in pending:
                    pending.append(eq)
        rws.append(rw)
    return rws


def huet_complete(eqs: list[Eq]) -> list[Rewrite]:
    # huet completion algorithm without marking
    # similar to previous
    # keep equations
    E, R = list(eqs), []
    while True:
        # reduce lhs, rhs according to R
        while E:
            eq = E.pop()
            lhs, rhs = reduce(eq.lhs, R), reduce(eq.rhs, R)
            if lhs == rhs:
                continue
            rw = Rewrite(lhs, rhs)

            # collapse old left sides and compose old right sides
            R1 = [rw]
            for rw1 in R:
                lhs1 = reduce(rw1.lhs, [rw])
                if lhs1 == rw1.lhs:
                    rhs1 = reduce(rw1.rhs, R + [rw])
                    if rw1.lhs != rhs1:
                        R1.append(Rewrite(rw1.lhs, rhs1))
                else:
                    E.append(Eq(lhs1, rw1.rhs))
            R = R1

        # create all critical pairs; marking would avoid redoing old pairs
        for i, rw in enumerate(R):
            for rw1 in R[: i + 1]:
                for (q, r), (q1, r1) in rw.lhs.overlaps_qr(rw1.lhs):
                    lhs = reduce(q * rw.rhs + r, R)
                    rhs = reduce(q1 * rw1.rhs + r1, R)
                    eq = Eq(lhs, rhs)
                    if lhs != rhs and eq not in E:
                        E.append(eq)
        if not E:
            return R


def huet_marked(eqs: list[Eq], goal: Optional[Eq] = None) -> list[Rewrite]:
    # the same as the above, but keep a bool in rewrite indicating if it is marked
    # the critical pair generation process picks one unwarked rules and generates critical pairs with all other rules, marking the rule after generating its critical pairs
    # if there are no unmarked rules, the algorithm terminates
    E, R = list(eqs), []
    while True:
        while E:
            eq = E.pop()
            lhs, rhs = reduce(eq.lhs, R), reduce(eq.rhs, R)
            if lhs == rhs:
                continue
            rw = Rewrite(lhs, rhs)

            R1 = [rw]
            for rw1 in R:
                lhs1 = reduce(rw1.lhs, [rw])
                if lhs1 == rw1.lhs:
                    rhs1 = reduce(rw1.rhs, R + [rw])
                    if rw1.lhs != rhs1:
                        R1.append(Rewrite(rw1.lhs, rhs1, rw1.mark))
                else:
                    E.append(Eq(lhs1, rw1.rhs))
            R = R1

            # Completion can be infinite even when the particular word problem
            # is already solved.  Equality after reduction by the current rules
            # is sound, so a goal-directed run may stop here.
            if goal is not None and reduce(goal.lhs, R) == reduce(goal.rhs, R):
                return R

        for rw in R:
            if not rw.mark:
                break
        else:
            return R

        rw.mark = True
        for rw1 in R:
            for (q, r), (q1, r1) in rw.lhs.overlaps_qr(rw1.lhs):
                lhs = reduce(q * rw.rhs + r, R)
                rhs = reduce(q1 * rw1.rhs + r1, R)
                eq = Eq(lhs, rhs)
                if lhs != rhs and eq not in E:
                    E.append(eq)


def _monom_quotient(big: MS, small: MS) -> Optional[MS]:
    """Return q when big = q*small, or None when small does not divide big."""
    q = list(big)
    for x in small:
        if x not in q:
            return None
        q.remove(x)
    return MS(q)


def rewrite_once(semi: Semi, eqs: list[Eq]):
    """Apply one equation in one monomial context, in either direction."""
    seen = set()
    for eq_index, eq in enumerate(eqs):
        for reverse, (old, new) in enumerate(((eq.lhs, eq.rhs), (eq.rhs, eq.lhs))):
            # Equations with an empty side need a separate bounded insertion rule.
            if not old.terms:
                continue
            for pm, _ in semi.terms:
                for om, _ in old.terms:
                    qm = _monom_quotient(pm, om)
                    if qm is None:
                        continue
                    context = Semi([qm])
                    rest = semi - context * old
                    if rest is None:
                        continue
                    result = rest + context * new
                    if result not in seen:
                        seen.add(result)
                        yield result, (eq_index, bool(reverse), qm)


def congruence_path(
    start: Semi,
    goal: Semi,
    eqs: list[Eq],
    *,
    max_degree: int,
    max_terms: int,
) -> Optional[list[Semi]]:
    """Bounded bidirectional search for a ground semiring-equational proof.

    `None` means no path was found within the bounds, not that the equation is
    false.  Each adjacent pair in a returned path differs by one use of an input
    equation in an additive and monomial multiplicative context.
    """

    def in_bounds(p: Semi) -> bool:
        return (
            all(len(m) <= max_degree for m, _ in p.terms)
            and sum(n for _, n in p.terms) <= max_terms
        )

    if start == goal:
        return [start]
    front = {start: None}
    back = {goal: None}
    front_queue, back_queue = deque([start]), deque([goal])

    while front_queue and back_queue:
        if len(front_queue) <= len(back_queue):
            here, there, queue = front, back, front_queue
        else:
            here, there, queue = back, front, back_queue

        for _ in range(len(queue)):
            semi = queue.popleft()
            for result, step in rewrite_once(semi, eqs):
                if not in_bounds(result) or result in here:
                    continue
                here[result] = (semi, step)
                if result in there:
                    left = []
                    p = result
                    while p is not None:
                        left.append(p)
                        parent = front[p]
                        p = None if parent is None else parent[0]
                    left.reverse()

                    right = []
                    p = result
                    while back[p] is not None:
                        p = back[p][0]
                        right.append(p)
                    return left + right
                queue.append(result)
    return None

File: _drafts/old/2026-09/semi.rs
use std::cmp::Ordering;
use std::collections::VecDeque;
use std::fmt;
use std::time::Instant;

/// Graded lexicographic order used for monomials and polynomials.
fn lenlex<T: Ord>(xs: &[T], ys: &[T]) -> Ordering {
    xs.len().cmp(&ys.len()).then_with(|| xs.cmp(ys))
}

#[derive(Clone, Debug, Eq, PartialEq)]
struct Monom(Vec<u32>);

impl Ord for Monom {
    fn cmp(&self, other: &Self) -> Ordering {
        lenlex(&self.0, &other.0)
    }
}

impl PartialOrd for Monom {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Monom {
    /// Multiset union of variables.
    fn mul(&self, other: &Self) -> Self {
        let (mut i, mut j) = (0, 0);
        let mut out = Vec::with_capacity(self.0.len() + other.0.len());
        while i < self.0.len() && j < other.0.len() {
            if self.0[i] <= other.0[j] {
                out.push(self.0[i]);
                i += 1;
            } else {
                out.push(other.0[j]);
                j += 1;
            }
        }
        out.extend_from_slice(&self.0[i..]);
        out.extend_from_slice(&other.0[j..]);
        Self(out)
    }

    /// Remove `other` as a submultiset.
    fn div(&self, other: &Self) -> Option<Self> {
        let (mut i, mut j) = (0, 0);
        let mut out = Vec::with_capacity(self.0.len() - other.0.len().min(self.0.len()));
        while i < self.0.len() {
            if j < other.0.len() && self.0[i] == other.0[j] {
                i += 1;
                j += 1;
            } else if j < other.0.len() && self.0[i] > other.0[j] {
                return None;
            } else {
                out.push(self.0[i]);
                i += 1;
            }
        }
        (j == other.0.len()).then_some(Self(out))
    }

    /// Componentwise maximum of variable multiplicities.
    fn lcm(&self, other: &Self) -> Self {
        let (mut i, mut j) = (0, 0);
        let mut out = Vec::with_capacity(self.0.len() + other.0.len());
        while i < self.0.len() || j < other.0.len() {
            let x = match (self.0.get(i), other.0.get(j)) {
                (Some(&x), Some(&y)) => x.min(y),
                (Some(&x), None) => x,
                (None, Some(&y)) => y,
                (None, None) => unreachable!(),
            };
            let i0 = i;
            let j0 = j;
            while self.0.get(i) == Some(&x) {
                i += 1;
            }
            while other.0.get(j) == Some(&x) {
                j += 1;
            }
            out.extend(std::iter::repeat_n(x, (i - i0).max(j - j0)));
        }
        Self(out)
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
struct Semi(Vec<Monom>);

type QR = (Semi, Semi);
type Overlap = (QR, QR);

impl Ord for Semi {
    fn cmp(&self, other: &Self) -> Ordering {
        lenlex(&self.0, &other.0)
    }
}

impl PartialOrd for Semi {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Semi {
    /// Build a polynomial in canonical sorted form.
    fn new(mut monoms: Vec<Monom>) -> Self {
        monoms.sort_unstable();
        Self(monoms)
    }

    /// Embed a generator.
    fn lit(x: u32) -> Self {
        Self(vec![Monom(vec![x])])
    }

    /// Embed a natural number as repeated empty monomials.
    fn of_int(n: usize) -> Self {
        Self(vec![Monom(Vec::new()); n])
    }

    /// Merge the two sorted multisets of monomials.
    fn add(&self, other: &Self) -> Self {
        let (mut i, mut j) = (0, 0);
        let mut out = Vec::with_capacity(self.0.len() + other.0.len());
        while i < self.0.len() && j < other.0.len() {
            if self.0[i] <= other.0[j] {
                out.push(self.0[i].clone());
                i += 1;
            } else {
                out.push(other.0[j].clone());
                j += 1;
            }
        }
        out.extend_from_slice(&self.0[i..]);
        out.extend_from_slice(&other.0[j..]);
        Self(out)
    }

    /// Distributive product of every pair of monomials.
    fn mul(&self, other: &Self) -> Self {
        let mut out = Vec::with_capacity(self.0.len() * other.0.len());
        for m in &self.0 {
            for n in &other.0 {
                out.push(m.mul(n));
            }
        }
        Self::new(out)
    }

    /// Multiply by one monomial and a natural coefficient.
    fn shift_scale(&self, q: &Monom, coeff: usize) -> Self {
        let mut out = Vec::with_capacity(self.0.len() * coeff);
        for m in &self.0 {
            let mq = m.mul(q);
            out.extend(std::iter::repeat_n(mq, coeff));
        }
        Self::new(out)
    }

    /// Natural subtraction, failing if `other` is not contained.
    fn sub(&self, other: &Self) -> Option<Self> {
        let (mut i, mut j) = (0, 0);
        let mut out = Vec::with_capacity(self.0.len().saturating_sub(other.0.len()));
        while j < other.0.len() {
            while i < self.0.len() && self.0[i] < other.0[j] {
                out.push(self.0[i].clone());
                i += 1;
            }
            if i == self.0.len() || self.0[i] != other.0[j] {
                return None;
            }
            i += 1;
            j += 1;
        }
        out.extend_from_slice(&self.0[i..]);
        Some(Self(out))
    }

    /// Coefficientwise maximum of two polynomials.
    fn union_max(&self, other: &Self) -> Self {
        let (mut i, mut j) = (0, 0);
        let mut out = Vec::with_capacity(self.0.len().max(other.0.len()));
        while i < self.0.len() || j < other.0.len() {
            match (self.0.get(i), other.0.get(j)) {
                (Some(m), Some(n)) if m == n => {
                    let (i0, j0) = (i, j);
                    while self.0.get(i) == Some(m) {
                        i += 1;
                    }
                    while other.0.get(j) == Some(n) {
                        j += 1;
                    }
                    out.extend(std::iter::repeat_n(m.clone(), (i - i0).max(j - j0)));
                }
                (Some(m), Some(n)) if m < n => {
                    out.push(m.clone());
                    i += 1;
                }
                (Some(_), Some(n)) => {
                    out.push(n.clone());
                    j += 1;
                }
                (Some(m), None) => {
                    out.push(m.clone());
                    i += 1;
                }
                (None, Some(n)) => {
                    out.push(n.clone());
                    j += 1;
                }
                (None, None) => break,
            }
        }
        Self(out)
    }

    /// Greedily find `q, r` with `self = q * other + r`.
    fn divrem(&self, other: &Self) -> (Self, Self) {
        assert!(!other.0.is_empty(), "division by zero");
        let mut q = Vec::new();
        let mut r = self.clone();
        let lm = other.0.last().unwrap();
        let mut i = r.0.len();
        while i > 0 {
            let Some(qm) = r.0[i - 1].div(lm) else {
                i -= 1;
                continue;
            };
            let term = other.shift_scale(&qm, 1);
            if let Some(r1) = r.sub(&term) {
                q.push(qm);
                r = r1;
                i = r.0.len();
            } else {
                i -= 1;
            }
        }
        (Self::new(q), r)
    }

    /// Return both `(quotient, remainder)` views of each nontrivial overlap.
    fn overlaps_qr(&self, other: &Self) -> Vec<Overlap> {
        let mut res = Vec::new();
        let mut seen = Vec::new();
        let mut i = 0;
        while i < self.0.len() {
            let mut i1 = i + 1;
            while i1 < self.0.len() && self.0[i1] == self.0[i] {
                i1 += 1;
            }
            let c1 = i1 - i;
            let mut j = 0;
            while j < other.0.len() {
                let mut j1 = j + 1;
                while j1 < other.0.len() && other.0[j1] == other.0[j] {
                    j1 += 1;
                }
                let c2 = j1 - j;
                let cm = self.0[i].lcm(&other.0[j]);
                let qm1 = cm.div(&self.0[i]).unwrap();
                let qm2 = cm.div(&other.0[j]).unwrap();
                let c = lcm(c1, c2);
                let q1 = Self(vec![qm1.clone(); c / c1]);
                let q2 = Self(vec![qm2.clone(); c / c2]);
                let t1 = self.shift_scale(&qm1, c / c1);
                let t2 = other.shift_scale(&qm2, c / c2);
                let ov = t1.union_max(&t2);
                if !seen.contains(&ov) {
                    seen.push(ov.clone());
                    res.push(((q1, ov.sub(&t1).unwrap()), (q2, ov.sub(&t2).unwrap())));
                }
                j = j1;
            }
            i = i1;
        }
        res
    }
}

impl fmt::Display for Semi {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.0.is_empty() {
            return write!(f, "0");
        }
        let mut i = 0;
        while i < self.0.len() {
            if i != 0 {
                write!(f, " + ")?;
            }
            let mut i1 = i + 1;
            while i1 < self.0.len() && self.0[i1] == self.0[i] {
                i1 += 1;
            }
            if i1 - i != 1 {
                write!(f, "{}*", i1 - i)?;
            }
            let m = &self.0[i].0;
            if m.is_empty() {
                write!(f, "1")?;
            } else {
                let mut j = 0;
                while j < m.len() {
                    let mut j1 = j + 1;
                    while j1 < m.len() && m[j1] == m[j] {
                        j1 += 1;
                    }
                    let x = char::from_u32(u32::from(b'a') + m[j]).unwrap_or('?');
                    write!(f, "{x}")?;
                    if j1 - j != 1 {
                        write!(f, "^{}", j1 - j)?;
                    }
                    j = j1;
                }
            }
            i = i1;
        }
        Ok(())
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
struct Rule {
    lhs: Semi,
    rhs: Semi,
    marked: bool,
}

impl Rule {
    /// Orient an equation from larger to smaller.
    fn new(mut lhs: Semi, mut rhs: Semi) -> Self {
        if lhs < rhs {
            std::mem::swap(&mut lhs, &mut rhs);
        }
        Self {
            lhs,
            rhs,
            marked: false,
        }
    }
}

/// Reduce to normal form, optionally followed by some extra rules.
fn reduce_with(mut semi: Semi, rewrites: &[Rule], extra: &[Rule]) -> Semi {
    'again: loop {
        for rw in rewrites.iter().chain(extra) {
            let (q, r) = semi.divrem(&rw.lhs);
            let res = q.mul(&rw.rhs).add(&r);
            if res != semi {
                semi = res;
                continue 'again;
            }
        }
        return semi;
    }
}

fn reduce(semi: Semi, rewrites: &[Rule]) -> Semi {
    reduce_with(semi, rewrites, &[])
}

/// Normalize all critical pairs between two rules.
fn critical_pairs<'a>(
    rw: &'a Rule,
    rw1: &'a Rule,
    rules: &'a [Rule],
) -> impl Iterator<Item = Rule> + 'a {
    rw.lhs
        .overlaps_qr(&rw1.lhs)
        .into_iter()
        .filter_map(move |((q, r), (q1, r1))| {
            let lhs = reduce(q.mul(&rw.rhs).add(&r), rules);
            let rhs = reduce(q1.mul(&rw1.rhs).add(&r1), rules);
            (lhs != rhs).then(|| Rule::new(lhs, rhs))
        })
}

fn gcd(mut a: usize, mut b: usize) -> usize {
    while b != 0 {
        (a, b) = (b, a % b);
    }
    a
}

fn lcm(a: usize, b: usize) -> usize {
    a / gcd(a, b) * b
}

/// Basic completion with a FIFO equation queue.
fn naive_complete(eqs: Vec<Rule>) -> (Vec<Rule>, usize) {
    let mut pending: VecDeque<_> = eqs.into();
    let mut rws = Vec::new();
    let mut popped = 0;
    while let Some(eq) = pending.pop_front() {
        popped += 1;
        let lhs = reduce(eq.lhs, &rws);
        let rhs = reduce(eq.rhs, &rws);
        if lhs == rhs {
            continue;
        }
        let rw = Rule::new(lhs, rhs);
        if rws.contains(&rw) {
            continue;
        }
        let new = rws.len();
        rws.push(rw);
        for old in 0..=new {
            for eq in critical_pairs(&rws[new], &rws[old], &rws) {
                if !pending.contains(&eq) {
                    pending.push_back(eq);
                }
            }
        }
    }
    (rws, popped)
}

/// Huet completion: collapse left sides and compose right sides.
fn huet_complete(mut eqs: Vec<Rule>) -> Vec<Rule> {
    let mut rws = Vec::new();
    loop {
        while let Some(eq) = eqs.pop() {
            let lhs = reduce(eq.lhs, &rws);
            let rhs = reduce(eq.rhs, &rws);
            if lhs == rhs {
                continue;
            }
            let rw = Rule::new(lhs, rhs);

            let mut rws1 = vec![rw.clone()];
            for rw1 in &rws {
                let lhs1 = reduce(rw1.lhs.clone(), std::slice::from_ref(&rw));
                if lhs1 == rw1.lhs {
                    let rhs1 = reduce_with(rw1.rhs.clone(), &rws, std::slice::from_ref(&rw));
                    if rw1.lhs != rhs1 {
                        rws1.push(Rule::new(rw1.lhs.clone(), rhs1));
                    }
                } else {
                    eqs.push(Rule::new(lhs1, rw1.rhs.clone()));
                }
            }
            rws = rws1;
        }

        for i in 0..rws.len() {
            for j in 0..=i {
                for eq in critical_pairs(&rws[i], &rws[j], &rws) {
                    if !eqs.contains(&eq) {
                        eqs.push(eq);
                    }
                }
            }
        }
        if eqs.is_empty() {
            return rws;
        }
    }
}

/// Huet completion, pairing each rule only after it becomes unmarked.
fn huet_marked(mut eqs: Vec<Rule>) -> Vec<Rule> {
    let mut rws: Vec<Rule> = Vec::new();
    loop {
        while let Some(eq) = eqs.pop() {
            let lhs = reduce(eq.lhs, &rws);
            let rhs = reduce(eq.rhs, &rws);
            if lhs == rhs {
                continue;
            }
            let rw = Rule::new(lhs, rhs);

            let mut rws1 = vec![rw.clone()];
            for rw1 in &rws {
                let lhs1 = reduce(rw1.lhs.clone(), std::slice::from_ref(&rw));
                if lhs1 == rw1.lhs {
                    let rhs1 = reduce_with(rw1.rhs.clone(), &rws, std::slice::from_ref(&rw));
                    if rw1.lhs != rhs1 {
                        let mut rw2 = Rule::new(rw1.lhs.clone(), rhs1);
                        rw2.marked = rw1.marked;
                        rws1.push(rw2);
                    }
                } else {
                    eqs.push(Rule::new(lhs1, rw1.rhs.clone()));
                }
            }
            rws = rws1;
        }

        let Some(i) = rws.iter().position(|rw| !rw.marked) else {
            return rws;
        };
        let rw = rws[i].clone();
        rws[i].marked = true;

        for rw1 in &rws {
            for eq in critical_pairs(&rw, rw1, &rws) {
                if !eqs.contains(&eq) {
                    eqs.push(eq);
                }
            }
        }
    }
}

/// Assert Buchberger's criterion for the finished system.
fn check_complete(rules: &[Rule]) {
    for rw in rules {
        for rw1 in rules {
            assert!(critical_pairs(rw, rw1, rules).next().is_none());
        }
    }
}

fn main() {
    let a = Semi::lit(0);
    let one = Semi::of_int(1);
    let seed = Rule::new(one.add(&a.mul(&a)), a);

    let start = Instant::now();
    let (naive, popped) = naive_complete(vec![seed.clone()]);
    println!(
        "naive: {} rules, {popped} equations in {:?}",
        naive.len(),
        start.elapsed()
    );

    let start = Instant::now();
    let rules = huet_complete(vec![seed.clone()]);
    println!("Huet: {} rules in {:?}", rules.len(), start.elapsed());
    for rw in &rules {
        println!("{} -> {}", rw.lhs, rw.rhs);
    }

    let start = Instant::now();
    let marked = huet_marked(vec![seed]);
    println!(
        "Huet marked: {} rules in {:?}",
        marked.len(),
        start.elapsed()
    );

    let start = Instant::now();
    check_complete(&naive);
    check_complete(&rules);
    check_complete(&marked);
    println!("all systems complete ({:?})", start.elapsed());

    assert_eq!(Semi::of_int(0).add(&Semi::of_int(2)), Semi::of_int(2));
}
